File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        arr = {}
        arr[None] = None
        cur = head 

        while cur: 
            copy = Node(cur.val)
            arr[cur] = copy
            cur = cur.next
        
        cur = head
        while cur: 
            arr[cur].next = arr[cur.next]
            arr[cur].random = arr[cur.random]
            cur = cur.next
        return arr[head]

File: Data Structures & Algorithms/reorder-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* fast = head; 
        ListNode* slow = head; 
        while( fast && fast->next){
            fast = fast->next->next;
            slow = slow->next; 
        }
        ListNode* l2 = slow->next;
        slow->next = nullptr;
        ListNode* prev = nullptr;
        ListNode* cur = l2; 
        while(cur){
            ListNode* temp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = temp;
        }
        ListNode* l1 = head;
        while(l1 && prev){
            ListNode* temp1 = l1->next;
            ListNode* temp2 = prev->next;
            l1->next = prev;
            prev->next = temp1;
            l1 = temp1;
            prev = temp2;
        }
    }
};
